### PR TITLE
Build: Trim UDK Size

### PR DESCRIPTION
**Pull request by dakanji**
Closes https://github.com/acidanthera/bugtracker/issues/1119

UDK size jumped up massively in moving to v0.6.0. 

Had a close look at things and this PR sets things back to a more reasonable size.

# 1: Default Full Recursive
- Gets all SubModules recursively and fetches the full history of each SubModule
- Total SubModules = 14
- Total Download Size = 1.15 Gb
- Speed = Very Very Slow
- OpenCorePkg Build = Builds Cleanly

# 2: Scripted Full Recursive
- Gets all SubModules recursively but only fetches the current head for each SubModule
- Total SubModules = 14
- Total Download Size = 0.72 Gb
- Speed = Quite Fast
- OpenCorePkg Build = Builds Cleanly

# 3: Scripted Recursive (Depth 2)
- Gets primary SubModules (Depth 1) with each able to get next level SubModules (Depth 2) and only fetches the current head for each SubModule
- Total SubModules = 14 (Same as Setup 2 which turns out to only go 2 levels)
- Total Download Size = 0.72 Gb . (Same as Setup 2 which turns out to only go 2 levels)
- Speed = Quite Fast
- OpenCorePkg Build = Builds Cleanly

# 4: Scripted Recursive (Depth 1)
- Gets primary SubModules (Depth 1) only and only fetches the current head for each SubModule
- Total SubModules = 6
- Total Download Size = 0.41 Gb
- Speed = Fast
- OpenCorePkg Build = Builds Cleanly

# 5: Scripted Recursive (Depth 0)
- Skips SubModules altogether
- Total SubModules = 0
- Total Download Size = 0.11 Gb
- Speed = Very Very Fast
- OpenCorePkg Build = Builds Falis

# 4a: Scripted Recursive (Depth 1 - Brotli Only)

Obviously Setup 4 is the most optimal as the deeper level UDK Submodules are not used by the OpenCorePkg build.
Looking at the build messages, it appeared that OpenCorePkg seemed to only require Brotli. Testing this gave:
- Gets primary Brotli SubModules Only and only fetches the current head for each
- Total SubModules = 2
- Total Download Size = 0.15 Gb
- Speed = Very Fast
- OpenCorePkg Build = Builds Cleanly

# 4b: Scripted Recursive (Depth 1 - Brotli Only ... Once)
Looking at the two Brotli SubModules, it turns out they are exactly the same and pulled form the same location.
- submodule.MdeModulePkg/Library/BrotliCustomDecompressLib/brotli.url=**https://github.com/google/brotli**
- submodule.BaseTools/Source/C/BrotliCompress/brotli.url=**https://github.com/google/brotli**

So, adjusted to download once and copy across to the other location:
- Gets primary Brotli SubModules Only and only fetches the current head for this andc opies across to other location
- Total SubModules = 1
- Total Download Size = 0.13 Gb
- Speed = Very Very Fast
- OpenCorePkg Build = Builds Cleanly

# Summary
This PR implements Scenario 4b upon which OpenCorePkg builds cleanly.
Will look for a rational amendment to the build instructions in the docs

### CODE REVIEW

**Comment by vit9696** [comment]
We use this not only for OpenCore, so it will not work quite right for us. For instance, we build OVMF packages. I am afraid your PR is not really a solution. You could try building other packages yourself and see that they fail this way.

**Comment by dakanji** [comment]
Can you give a list of packages outside OC?
Even if the Brotli only version does not meet the requirements of these, the `Scripted Full Recursive` gives exactly the same fileset as the current version but at a reduced size and is a lot faster for instance.

**Comment by vit9696** [comment]
Well, I cannot tell you much about the closed-source projects, but you could build OVMF for a test:


```
#!/bin/bash

cd "$(dirname "$0")" || exit 1
if [ "$ARCHS" = "" ]; then
  ARCHS=X64
fi

package() {
  if [ ! -d "$1"/../FV ]; then
    echo "Missing FV directory $1/../FV"
    exit 1
  fi

  if [ ! -f "$1"/../FV/OVMF.fd ]; then
    echo "Missing OVMF.fd directory in $1/../FV"
    exit 1
  fi

  cp "$1"/../FV/OVMF.fd "Binaries/OVMF-${ARCHS}.fd" || exit 1
}

if [ "$ARCHS" = "IA32" ]; then
  SELFPKG=OvmfPkgIa32
  RELPKG=OvmfIa32
elif [ "$ARCHS" = "X64" ]; then
  SELFPKG=OvmfPkgX64
  RELPKG=OvmfX64
else
  echo "Unsupported architecture $ARCHS"
fi
SELFPKG_DIR=OvmfPkg
NO_ARCHIVES=1
TARGETS=RELEASE
RTARGETS=RELEASE

export ARCHS
export SELFPKG
export RELPKG
export SELFPKG_DIR
export NO_ARCHIVES
export TARGETS
export RTARGETS

if [ "$(uname)" = "Linux" ]; then
  TOOLCHAINS=GCC5
  export TOOLCHAINS
fi

if [ ! -f "mtoc-mac64.sha256" ]; then
  curl -LO https://github.com/acidanthera/ocbuild/raw/master/external/mtoc-mac64.sha256 || exit 1
fi
MTOC_HASH=$(cat mtoc-mac64.sha256)
export MTOC_HASH

if [ ! -f "efibuild.sh" ]; then
  curl -LO https://raw.githubusercontent.com/acidanthera/ocbuild/master/efibuild.sh || exit 1
fi
. efibuild.sh

**Comment by dakanji** [comment]
> Well, I cannot tell you much about the closed-source projects

Noted.

Will amend to the version that returns the same file set as the current full history setup.
I suppose that the docs for **OpenCore** can outline something that achieves this version

**Comment by vit9696** [comment]
I think what we can do is to pass the submodules from the OpenCore build file to efibuild. When unspecified it could build all, otherwise only some. But maybe try to build OVMF first and see how it goes.

**Comment by dakanji** [comment]
Probably easier to go for Setup 2 which replicates the submodules as they currently are as it is already smaller and a lot faster.

A more nuanced implementation can then come after.

**Comment by vit9696** [comment]
Hmm, but does not the current implementation clone submodules without history? Or is that a git limitation?

**Comment by vit9696** [comment]
I think you could try `git submodule update --init --recommend-shallow`.

**Comment by dakanji** [comment]
Can you reopen so I can push change?

**Comment by dakanji** [comment]
> Hmm, but does not the current implementation clone submodules without history? Or is that a git limitation?
Current Implementation fetches the entire history. Openssl is huge for instance and that's before it gets its own submodules such as boringssl which is also huge.

> I think you could try git submodule update --init --recommend-shallow instead if --recursive.
Every native implementation comes with ifs and buts. You basically can't guaranty it will work on every git flavour you can reasonably expect out in the wild.

What this does is use a recursive function that reads the .gitmodule file and creates the repo with depth=1. It then reads the file in that new one and does the same things and gain and again.

I have the changes ready and just waiting for reopen.

**Comment by vit9696** [comment]
Well, this function is supported by git and does not depend on the repository structure. I believe size-wise it is essentially 4, and for all kinds of reasons I would strongly prefer using it over custom scripts as we update EDK II with upstream with the time, and we definitely want to make least changes when this happens.

If you would like to continue working on the changes within this PR, I reopened it.

**Comment by dakanji** [comment]
> I would strongly prefer using it over custom scripts as we update EDK II with upstream with the time, and we definitely want to make least changes when this happens.

Get your point but please note that this operates exactly as Git would. I.E., it reads the available .gitmodule files using internal git function and only uses what is there. If any .gitmodule files changes tomorrow, the script function will fetch whatever is there. 

Basically works with any changes along the path of the submodules and doesn't create a situation where you need to go back and tinker with it.

I made it slightly less efficient (not copying the repeated brotli but fetching it again) to make 100% sure that should there be a change, it will follow that change. Not really needed but playing safe to give comfort.

**Comment by dakanji** [comment]
See amended

**Comment by vit9696** [comment]
Yes, but your function does far less error checking, may have issues with directories containing spaces and so on. Why would you use a lengthy solution when there is a one-liner supported by the vendor? Is there an operating system where it does not work? Is there any performance difference?

**Comment by dakanji** [comment]
> your function does far less error checking

Not sure what error checking exactly git does. AFAIK, it reads the `gitmodules` file and follows what is there. Which is what this does. Minor issue anyway. Your main point is the simplicity and vendor support.

> may have issues with directories containing spaces and so on. 

No issues with that. All the paths are properly captured. Shellcheck in the CI would have flagged such issues if existing 

> Why would you use a lengthy solution when there is a one-liner supported by the vendor? 

- The one-liner does not seem to give the same number of submodules as the current implementation. You mentioned that you needed all previously. I think it may depend on some flags being present in the upstream gitmodules file setup.
Screen Shot 2020-08-25 at 7 55 49 AM
- If you don't need whatever the two missing ones are (unless I counted wrong when I counted 14), then seems ok.

> Is there an operating system where it does not work? 

- Basically git version dependent (Seems to have come in with Git 2.9.0 in 2016). I think included with HiSierra onward on Macs which I believe came with Git 2.14.

> Is there any performance difference?

- A bit slower on my unit but definitely a lot better than the current implementation.

Anyway, will leave this for now as need to run. Will check in later.

**Comment by vit9696** [comment]
As for error checking, I meant that your script may not exit properly if one of the commands fails, while with the one-liner it is more reliable.

As for cloning less submodules by one liner, I believe this is the point. Repository vendor can provide less submodules for a shallow copy when needed. Which is kind of useful.

As for High Sierra, I am not positive we still support Sierra or older.

To be honest, need somebody with git knowledge to comment on this.

**Comment by dakanji** [comment]
> As for cloning less submodules by one liner, I believe this is the point ... Which is kind of useful.

Don't fully understand the point being made but assume you prefer one-liner and will change to this when back at machine.

**Comment by dakanji** [comment]
Changed to one-liner

File: efibuild.sh [hunk 1]
```diff
@@ -17,7 +17,14 @@ prompt() {
 
 updaterepo() {
   if [ ! -d "$2" ]; then
-    git clone --recursive "$1" -b "$3" --depth=1 "$2" || exit 1
+      if [ "$2" == "UDK" ] ; then
+          git clone "$1" -b "$3" --depth=1 "$2" || exit 1
+          pushd "$2" >/dev/null || exit 1
+          git submodule update --init --recommend-shallow
```
Comment (vit9696) [inline efibuild.sh:23]: || exit 1 is missing

File: efibuild.sh [hunk 2]
```diff
@@ -17,7 +17,14 @@ prompt() {
 
 updaterepo() {
   if [ ! -d "$2" ]; then
-    git clone --recursive "$1" -b "$3" --depth=1 "$2" || exit 1
+      if [ "$2" == "UDK" ] ; then
```
Comment (vit9696) [inline efibuild.sh:20]: Why do we make UDK any special? This can be done for any repository, no?

Comment (dakanji) [reply efibuild.sh:20]: It can actually ... only did like that to minimise scope of impact of changes when things were up in the air.

Comment (vit9696) [reply efibuild.sh:20]: Please change that then, and also make sure to maintain the same indentation as before.

Comment (dakanji) [reply efibuild.sh]: Manually copied existing indentations but better if you add `.editorconfig` files to the repos so that your style can be picked up by IDEs.

PS. moved the update line so that it applies to new cloning and updates.
